### PR TITLE
ECKIT-619 Enable PUT method on EasyCURL

### PR DESCRIPTION
**Pull request by marcosbento**
This pull request provides a `readCallback` to enable uploading the HTTP PUT request body, and updates the `EasyCURL::PUT` to effectively process the PUT method request.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/eckit?pullRequest=81) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/eckit?pullRequest=81) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/eckit?pullRequest=81) it.</sub>

**Review by tlmquintino — COMMENTED** [review]
Looks good. 
Can we add a test please? or is there no tests for the curl?
I think we need to add tests that then could use our nexus servers to test the code both for gets and puts.

**Comment by marcosbento** [comment]
> Looks good. Can we add a test please? or is there no tests for the curl? I think we need to add tests that then could use our nexus servers to test the code both for gets and puts.

Currently there are no tests for CURL, but I'll definitely have a look at your suggestion regarding using our Nexus servers as target end points.

File: tests/io/test_easycurl.cc [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+CASE("EasyCURL GET (REST API, Successful operation)") {
+    EasyCURLHeaders headers;
+    headers["content-type"] = "application/json";
+
+    auto curl = EasyCURL();
+    curl.headers(headers);
+
+    auto response = curl.GET("https://get.ecmwf.int/service/rest/v1/assets?repository=test-data");
+    EXPECT(response.code() == 200);
```
Comment (simondsmart) [inline tests/io/test_easycurl.cc:62]: Can we have some GET tests that fail (i.e. not just code == 200)

File: tests/io/test_easycurl.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+CASE("EasyCURL PUT (REST API, Insufficient permissions)") {
+    EasyCURLHeaders headers;
+    headers["content-type"] = "application/json";
+
+    auto curl = EasyCURL();
+    curl.headers(headers);
+
+    auto response = curl.PUT("https://get.ecmwf.int/service/rest/v1/security/anonymous", "");
+    EXPECT(response.code() == 403);
```
Comment (simondsmart) [inline tests/io/test_easycurl.cc:86]: Any chance of a successful PUT request? Nothing here demonstrates that the data is being passed through properly.

Ideally a PUT request that can re-retrieve the data, and compare.

**Review by simondsmart — APPROVED** [review]
This looks good overall.

Ideally I would like to see a GET/PUT test that fails/succeeds as appropriate.

File: tests/io/test_easycurl.cc (hunk 2 above)
Comment (marcosbento) [reply tests/io/test_easycurl.cc:86]: I'm consulting with @iainrussell to see if there is any way to make a PUT request on our Nexus, without needed authentication.

If you have any idea, let us know.

**Comment by marcosbento** [comment]
Hi @simondsmart , @tlmquintino

I've added a more elaborate test (covering GET, POST, PUT and DELETE methods) , based on a mock REST API implemented in Python+Flask. Can you please have another look?

**Comment by marcosbento** [comment]
Hi @simondsmart , @tlmquintino

Are you happy to merge this PR now?

**Pull request merged**